### ISSUE TITLE
Not working with iPad2,4, iPad2,5, iPad2,6, iPad2,7

### ISSUE THREAD

**Issue by DevFlorian** [LukeZGD/32bit-OTA-Downgrader#7]
I'm getting this error:

Detected device has no baseband
Version:  XXXX - 152
Libipatcher Version: XXXX - 44
Odysseus Support: yes
[INFO] 32bit device detected
futurerestore init done
reading ticket 4394859152137_iPad2,5_8.4.1-12H321_XXXX.shsh2 done

WARNING: user specified not to flash a baseband. This can make the restore fail if the device needs a baseband!
if you added this flag by mistake you can press CTRL-C now to cancel
continuing restore in 5 4 3 2 1 
Found device in DFU mode
requesting to get into pwnRecovery later
Found device in DFU mode
Identified device as p105ap, iPad2,5
Extracting BuildManifest from IPSW
Product Version: 8.4.1
Product Build: 12H321 Major: 12
Device supports Image4: false
checking APTicket to be valid for this restore...
Verified ECID in APTicket matches device ECID
[WARNING] skipping ramdisk hash check, since device is in pwnDFU according to user
Variant: Customer Erase Install (IPSW)
This restore will erase your device data.
Extracting iBSS.p105.RELEASE.dfu...
terminate called after throwing an instance of 'libipatcher::exception'
  what():  std::exception
./restore.sh: line 145:  6233 Aborted                 (core dumped) sudo env "LD_PRELOAD=libcurl.so.3" tools/futurerestore_$platform -t $SHSH --no-baseband --use-pwndfu ${IPSW}.ipsw

futurerestore done!
If futurerestore failed to download baseband or for some reason, you can choose to retry
Retry? (y/n)
y

**Comment by DevFlorian** [comment]
I used an older version. Current version throws same error in line 136.

**Comment by LukeZGD** [comment]
What distro are you using? Also are you in live USB or full install? I don't support VMs anymore since they can have trouble working

**Comment by Morph1984** [comment]
@LukeZGD I was helping someone downgrade to 8.4.1 on the iPad 2,5, it failed for them in the futurerestore step. I had them follow this guide to resolve it https://www.reddit.com/r/jailbreak/comments/7sd9qb/tutorialhow_to_fix_libcabidylib_terminating_with/

```
WARNING: user specified not to flash a baseband. This can make the restore fail if the device needs a baseband!
if you added this flag by mistake you can press CTRL-C now to cancel
continuing restore in 5 4 3 2 1 
Found device in DFU mode
requesting to get into pwnRecovery later
Found device in DFU mode
Identified device as p105ap, iPad2,5
Extracting BuildManifest from IPSW
Product Version: 8.4.1
Product Build: 12H321 Major: 12
Device supports Image4: false
checking APTicket to be valid for this restore...
Verified ECID in APTicket matches device ECID
[WARNING] skipping ramdisk hash check, since device is in pwnDFU according to user
Variant: Customer Erase Install (IPSW)
This restore will erase your device data.
Extracting iBSS.p105.RELEASE.dfu...
terminate called after throwing an instance of 'libipatcher::exception'
  what():  std::exception
Aborted
```

**Comment by Morph1984** [comment]
I told him to run this command
```
mkdir "~/firmware/iPad2,5/12H321"
```

Place the appropriate index.html in the 12H321 folder, and then run
```
sudo python -m https.server 80
```
(python3)
in a separate terminal window, and re-run the script

**Comment by LukeZGD** [comment]
> @LukeZGD I was helping someone downgrade to 8.4.1 on the iPad 2,5, it failed for them in the futurerestore step. I had them follow this guide to resolve it https://www.reddit.com/r/jailbreak/comments/7sd9qb/tutorialhow_to_fix_libcabidylib_terminating_with/
> 
> ```
> WARNING: user specified not to flash a baseband. This can make the restore fail if the device needs a baseband!
> if you added this flag by mistake you can press CTRL-C now to cancel
> continuing restore in 5 4 3 2 1 
> Found device in DFU mode
> requesting to get into pwnRecovery later
> Found device in DFU mode
> Identified device as p105ap, iPad2,5
> Extracting BuildManifest from IPSW
> Product Version: 8.4.1
> Product Build: 12H321 Major: 12
> Device supports Image4: false
> checking APTicket to be valid for this restore...
> Verified ECID in APTicket matches device ECID
> [WARNING] skipping ramdisk hash check, since device is in pwnDFU according to user
> Variant: Customer Erase Install (IPSW)
> This restore will erase your device data.
> Extracting iBSS.p105.RELEASE.dfu...
> terminate called after throwing an instance of 'libipatcher::exception'
>   what():  std::exception
> Aborted
> ```

Thanks for sharing this! So it appears that futurerestore fails because of a missing KBAG key for 12H321 in [ipsw.me API](https://firmware-keys.ipsw.me/firmware/iPad2,5/12H321), and not just iPad2,5, but 2,4, 2,6, 2,7 are affected as well.. I'll try getting the KBAG keys for those, then apply some fixes in the script later

**Comment by LukeZGD** [comment]
I have added the fixes for iPad2,5 now, please try again on the latest commit and see if it works. I'm currently getting the KBAG keys for other devices, iPod5,1 seems to be affected as well

**Comment by DevFlorian** [comment]
Now it stops on:
Preparing for futurerestore...
Serving HTTP on 0.0.0.0 port 80 ...

I've been waiting for a couple of minutues.

I'm on Lubuntu 16.04 Live USB

**Comment by LukeZGD** [comment]
> Now it stops on:
> Preparing for futurerestore...
> Serving HTTP on 0.0.0.0 port 80 ...
> 
> I've been waiting for a couple of minutues.
> 
> I'm on Lubuntu 16.04 Live USB

That's weird, I'm pretty sure that's fixed on 71013168e6ad042f2d13080ca390a955dbf8e3eb, it seems I should test further

**Comment by DevFlorian** [comment]
Now it‘s working. Somehow i downloaded an old version as zip.
